addon/lib/Feature.jsm

```diff
@@ -1,140 +1,660 @@
+      this._log.debug("_performMeasurement - control group, bailing out");
+      return;
+    }
+
+    // Are we online? Do we have network connectivity?
+    if (this._offline()) {
+      this._log.debug("_performMeasurement - offline");
+      return;
+    }
+
+    // Check if enough time passed since the last download.
+    const lastMeasurementDate = getDateFromPref(STUDY_PREFS.LastMeasurement);
+    if (lastMeasurementDate &&
+        Math.abs(lastMeasurementDate - Date.now()) < this._delayBetweenMeasurements) {
+      this._log.debug(`_performMeasurement - skipping idle (last measurement ${lastMeasurementDate})`);
+      return;
+    }
+
+    // Bail out if a measurement is already in progress.
+    if (this._isMeasuring) {
+      this._log.warn("_performMeasurement - measurement in progress");
+      return;
+    }
+
+    // Trigger the request.
+    let endpointUrl =
+      Services.prefs.getCharPref(STUDY_PREFS.Endpoint.name, `${STUDY_PREFS.Endpoint.defaultValue}`);
+    // Append a random number to the request to bypass the cache.
+    endpointUrl = endpointUrl + "?" + (new Date()).getTime();
+
+    // Generate the request and load the script in the browser.
+    this._isMeasuring = true;
+    const url = this._triggerRequest(endpointUrl);
+    this._browser.setAttribute("src", url);
+
+    // Set the last measurement date: we don't care if it was a failure or success,
+    // as we don't want to measure more frequently than required.
+    Services.prefs.setCharPref(STUDY_PREFS.LastMeasurement.name, `${Date.now()}`);
+  }
+
+  /**
+   * Generate a custom ping and send it along with our data,
+   *
+   * @param {Object} data The full or partial data to send with the ping.
+   * @param {String} reason The reason for sending this ping, i.e. "progress"
+   *                 or "aborted".
+   */
+  async _generateAndSendPing(data, reason) {
+    // Send the measurement data.
+    let payload = {
+      reason,
+      type: data.type || "",
+      latency: data.latency,
+      fileSize: data.fileSize,
+      connType: data.connType,
+      goodput: [],
+    };
+
+    let computeGoodput = (elapsedTimeMs, downloadedInBytes) => {
+      const BYTES_PER_MEBIBYTE = 1048576;
+      const downloadedMebibytes = downloadedInBytes / BYTES_PER_MEBIBYTE;
+      const elapsedSeconds = elapsedTimeMs / 1000.0;
+      return (downloadedMebibytes * 8) / elapsedSeconds;
+    };
+
+    const checkpoints = data.transferCheckpoints || [];
+    for (let t of checkpoints) {
+      const elapsedMs = t[0];
+      payload.goodput.push([elapsedMs, computeGoodput(elapsedMs, t[1])]);
+    }
+
+    // Unfortunately, we can't simply use |this._studyUtils.telemetry| as
+    // it requires our data to be in a specific format to be serialized to
+    // the experiments parquet dataset.
+    try {
+      if (!this._studyUtils.telemetryConfig.send) {
+        return;
+      }
+    } catch (ex) {
+      this._log.error("_generateAndSendPing", ex);
```

Comment:
What's the failure case you're expecting to find here?

Reply:
I'm not sure if there's anything other than "running without SHIELDUtils being enable". It's just to be on the safe side, since the API is designed to throw.

Reply:
I still don't understand. The code in the try statement is

```js
if (!this._studyUtils.telemetryConfig.send) {
  return;
}
```

and then it goes on to not use study utils at all, instead calling out to Telemetry manually. Given that studyUtils is a vendored library that is tied to this code, how would you be "running without  SHIELDUtils being enabled", and how does that affect the code later in this function?

Reply:
The try/catch block is needed due to [this line](https://github.com/mozilla/shield-studies-addon-utils/blob/9f94a24aadcaa4e32aa1d54b6b5a5703392bb494/src/StudyUtils.in.jsm#L256): the API is designed to throw if it gets called without the setup being completed first. I know we're supposed to run the setup first, and that's happening, but since we're working in `chrome` I don't really want to leave anything in a broken state in case of weird edge cases.

Reply:
Ok, that makes sense how an error would be thrown (I forget that fallible getters are a thing sometime, sorry). However, it still leaves my other question. After the try/catch block, the code is

```
const options = {addClientId: true, addEnvironment: true};
await TelemetryController.submitExternalPing("shield-icq-v1", payload, options);
```

This doesn't use the study utils at all, as far as I can tell. Is there some benefit to checking for checking for the telemetryConfig method directly? Or would calling `.throwIfNotSetup()` directedly be equally useful?

Sorry for harping on this. It seems like a minor detail, but I don't feel comfortable signing off if I don't understand whats the reasoning behind the code is, and this is very confusing to me.

Reply:
No problem, it's ok to discuss this :) The code after the try-catch generates a custom ping with our payload. However, there might be cases for which we don't even want the payload to be submitted to Telemetry (e.g. SHIELD being disabled while Telemetry is enabled).

To make sure we submit the payload with the same behaviour as `ShieldUtils`, I'm checking the same thing [they are checking](https://github.com/mozilla/shield-studies-addon-utils/blob/9f94a24aadcaa4e32aa1d54b6b5a5703392bb494/src/StudyUtils.in.jsm#L397) before submitting the ping.